From: Thomas Renninger <trenn@suse.de>
Subject: XEN: Implement ioremap_page_range for XEN used by APEI
References: fate#311701,fate#311703,fate#311704,fate#311771,fate#311706,fate#311772,bnc#697859
Patch-Mainline: no

This part was copied from our master branch XEN patches:
patches.xen/xen3-patch-2.6.38
to compile again with latest APEI enhancements:
patches.drivers/001-apei_infrastructure.patch

Signed-off-by: Thomas Renninger <trenn@suse.de>

---
 arch/x86/mm/ioremap-xen.c |   10 ++++++++++
 1 file changed, 10 insertions(+)

Index: linux-2.6.32-SLE11-SP2/arch/x86/mm/ioremap-xen.c
===================================================================
--- linux-2.6.32-SLE11-SP2.orig/arch/x86/mm/ioremap-xen.c
+++ linux-2.6.32-SLE11-SP2/arch/x86/mm/ioremap-xen.c
@@ -197,6 +197,16 @@ int page_is_ram(unsigned long pagenr)
 	return 0;
 }
 
+#ifdef CONFIG_MODULES
+/*
+ * Force the implementation of ioremap_page_range() to be pulled in from
+ * lib/lib.a even if there is no other reference from the core kernel to it
+ * (native uses it in __ioremap_caller()), so that it gets exported.
+ */
+static void *const __section(.discard.ioremap) __used
+_ioremap_page_range = ioremap_page_range;
+#endif
+
 /*
  * Fix up the linear direct mapping of the kernel to avoid cache attribute
  * conflicts.
